launcher_go/v2/config/config.go

```diff
@@ -0,0 +1,209 @@
+	}
+	matched, _ := regexp.MatchString("[[:upper:]/ !@#$%^&*()+~`=]", configName)
+	if matched {
+		msg := "ERROR: Config name '" + configName + "' must not contain upper case characters, spaces or special characters. Correct config name and rerun."
+		fmt.Println(msg)
+		return nil, errors.New(msg)
+	}
+
+	config_filename := string(strings.TrimRight(dir, "/") + "/" + config.Name + ".yml")
+	content, err := os.ReadFile(config_filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			fmt.Println("config file does not exist: " + config_filename)
+		}
+		return nil, err
+	}
+	baseConfig := &Config{}
+
+	if err := yaml.Unmarshal(content, baseConfig); err != nil {
+		return nil, err
+	}
+
+	if includeTemplates {
+		for _, t := range baseConfig.Templates {
+			if err := config.loadTemplate(templatesDir, t); err != nil {
+				return nil, err
+			}
+		}
+	}
+	if err := mergo.Merge(config, baseConfig, mergo.WithOverride); err != nil {
+		return nil, err
+	}
+	config.rawYaml = append(config.rawYaml, string(content[:]))
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range config.Labels {
+		val := strings.ReplaceAll(v, "{{config}}", config.Name)
+		config.Labels[k] = val
+	}
+
+	for k, v := range config.Env {
+		val := strings.ReplaceAll(v, "{{config}}", config.Name)
+		config.Env[k] = val
+	}
+
+	return config, nil
+}
+
+func (config *Config) Yaml() string {
+	return strings.Join(config.rawYaml, "_FILE_SEPERATOR_")
+}
+
+func (config *Config) Dockerfile(pupsArgs string, bakeEnv bool) string {
+	builder := strings.Builder{}
+	builder.WriteString("ARG dockerfile_from_image=" + config.Base_Image + "\n")
+	builder.WriteString("FROM ${dockerfile_from_image}\n")
+	builder.WriteString(config.DockerfileArgs() + "\n")
+	if bakeEnv {
+		builder.WriteString(config.DockerfileEnvs() + "\n")
+	}
+	builder.WriteString(config.DockerfileExpose() + "\n")
+	builder.WriteString("COPY config.yaml /temp-config.yaml\n")
+	builder.WriteString("RUN " +
+		"cat /temp-config.yaml | /usr/local/bin/pups " + pupsArgs + " --stdin " +
+		"&& rm /temp-config.yaml\n")
+	builder.WriteString("CMD [\"" + config.BootCommand() + "\"]")
+	return builder.String()
+}
+
+func (config *Config) WriteYamlConfig(dir string) error {
+	file := strings.TrimRight(dir, "/") + "/config.yaml"
+	if err := os.WriteFile(file, []byte(config.Yaml()), 0660); err != nil {
+		return errors.New("error writing config file " + file)
+	}
+	return nil
+}
+
+func (config *Config) BootCommand() string {
+	if len(config.Boot_Command) > 0 {
+		return config.Boot_Command
+	} else if config.No_Boot_Command {
+		return ""
+	} else {
+		return "/sbin/boot"
+	}
+}
+
+func (config *Config) EnvArray(includeKnownSecrets bool) []string {
+	envs := []string{}
+	for k, v := range config.Env {
+		if !includeKnownSecrets && slices.Contains(utils.KnownSecrets, k) {
+			continue
+		}
+		envs = append(envs, k+"="+v)
+	}
+	slices.Sort(envs)
+	return envs
+}
+
+func (config *Config) DockerfileEnvs() string {
```

Comment:
```suggestion
func (config *Config) dockerfileEnvs() string {
```

This seems to be an internal function and does not need to be exported so the function name should be lower case.